File: projects/five-cycle-double-cover/scratch/d5-simultaneous-external-frontier-20260731/check_simultaneous.cpp
#define main typed_cap_original_main
#include "../d5-typed-cap-double-star-frontier-20260731/audit_d5_typed_cap_ports.cpp"
#undef main

#include <algorithm>
#include <iostream>
#include <stdexcept>
#include <string>

// Complete exact census of the all-flow simultaneous property.  For every
// proper rooted interface (z,r), some one D5 flow must have external root
// factors whose physical two-port pairs cover all three ports of z.

int main() {
    std::string row;
    long long graph_count = 0;
    long long flow_count = 0;
    long long interface_count = 0;
    long long witnessed_count = 0;
    while (std::cin >> row) {
        if (!row.empty() && row[0] == '>') continue;
        Graph graph = decode_graph6(row);
        Audit audit(graph);
        const auto all = audit.flows();
        flow_count += static_cast<long long>(all.size());

        const int m = static_cast<int>(graph.edges.size());
        std::vector<unsigned char> witnessed(graph.n * m, 0);
        std::vector<unsigned char> proper(graph.n * m, 0);
        for (int z = 0; z < graph.n; ++z) {
            for (int root = 0; root < m; ++root) {
                if (graph.edges[root].first == z || graph.edges[root].second == z) continue;
                proper[z * m + root] = 1;
                ++interface_count;
            }
        }

        for (const State& state : all) {
            std::vector<unsigned char> external_ports(graph.n * m, 0);
            for (unsigned char pair : LABELS) {
                for (const auto& component : audit.components(state, pair)) {
                    std::vector<unsigned char> in(m, 0);
                    for (int edge : component) in[edge] = 1;
                    for (int z = 0; z < graph.n; ++z) {
                        std::vector<int> slots;
                        int inactive = -1;
                        for (int slot = 0; slot < 3; ++slot) {
                            const int edge = graph.incidence[z][slot];
                            if (in[edge]) slots.push_back(slot);
                            else inactive = slot;
                        }
                        if (slots.size() != 2) continue;
                        const unsigned char inactive_label = static_cast<unsigned char>(
                            state[graph.incidence[z][inactive]]);
                        if (pair == inactive_label) continue;
                        if (pair & inactive_label)
                            throw std::runtime_error("invalid inactive factor relation");
                        const unsigned char port_bits = static_cast<unsigned char>(
                            (1 << slots[0]) | (1 << slots[1]));
                        for (int root : component) {
                            if (proper[z * m + root])
                                external_ports[z * m + root] |= port_bits;
                        }
                    }
                }
            }
            for (int index = 0; index < graph.n * m; ++index) {
                if (external_ports[index] == 7) witnessed[index] = 1;
            }
        }

        for (int index = 0; index < graph.n * m; ++index) {
            if (!proper[index]) continue;
            if (witnessed[index]) {
                ++witnessed_count;
            } else {
                const int z = index / m;
                const int root = index % m;
                std::cout << "FAIL graph=" << graph_count
                          << " z=" << z << " root=" << root << "\n";
            }
        }
        ++graph_count;
    }
    std::cout << "SIMULTANEOUS graphs=" << graph_count
              << " flows=" << flow_count
              << " interfaces=" << interface_count
              << " witnessed=" << witnessed_count << "\n";
    if (witnessed_count != interface_count) return 2;
    std::cout << "PASS\n";
    return 0;
}

File: projects/five-cycle-double-cover/scratch/d5-simultaneous-external-frontier-20260731/check_simultaneous.py
#!/usr/bin/env python3
"""Independent small-order replay of one-flow external port coverage."""

from __future__ import annotations

from collections import Counter
import os
from pathlib import Path
import subprocess
import sys


HERE = Path(__file__).resolve().parent
TYPED = HERE.parent / "d5-typed-cap-double-star-frontier-20260731"
sys.path.insert(0, str(TYPED))
import verify_small_and_algebra as base  # noqa: E402


def graph6_rows(order: int) -> list[str]:
    geng = os.environ.get("GENG", "/opt/homebrew/bin/geng")
    run = subprocess.run(
        [geng, "-Cq", "-d3", "-D3", str(order)],
        check=True,
        text=True,
        capture_output=True,
    )
    return [row for row in run.stdout.splitlines() if row and not row.startswith(">")]


def fixed_external_bits(n, edges, incidence, labels):
    result = {
        (z, root): 0
        for z in range(n)
        for root, endpoints in enumerate(edges)
        if z not in endpoints
    }
    for pair in base.LABELS:
        for component in base.active_components(labels, pair, edges, incidence):
            component_set = set(component)
            for z in range(n):
                slots = tuple(
                    slot for slot, edge in enumerate(incidence[z])
                    if edge in component_set
                )
                if len(slots) != 2:
                    continue
                inactive = next(slot for slot in range(3) if slot not in slots)
                third = labels[incidence[z][inactive]]
                if pair == third:
                    continue
                assert not (pair & third)
                port_bits = sum(1 << slot for slot in slots)
                for root in component:
                    if z not in edges[root]:
                        result[z, root] |= port_bits
    return result


def main() -> None:
    expected = {
        4: (1, 18, 12),
        6: (2, 138, 72),
        8: (5, 1464, 360),
        10: (18, 18012, 2160),
        12: (81, 309678, 14580),
    }
    for order, (expected_graphs, expected_flows, expected_interfaces) in expected.items():
        graph_count = flow_count = interface_count = 0
        maxima: list[int] = []
        for row in graph6_rows(order):
            n, edges, incidence = base.decode_graph6(row)
            assert n == order
            local = {
                (z, root): 0
                for z in range(n)
                for root, endpoints in enumerate(edges)
                if z not in endpoints
            }
            for labels in base.enumerate_representatives(edges, incidence):
                flow_count += 1
                values = fixed_external_bits(n, edges, incidence, labels)
                for key, bits in values.items():
                    local[key] = max(local[key], bits.bit_count())
            maxima.extend(local.values())
            graph_count += 1
        interface_count = len(maxima)
        counts = Counter(maxima)
        assert (graph_count, flow_count, interface_count) == (
            expected_graphs, expected_flows, expected_interfaces
        )
        assert counts == {3: interface_count}
        print(
            f"PYTHON_SIMULTANEOUS order={order} graphs={graph_count} "
            f"representative_flows={flow_count} interfaces={interface_count} "
            "PASS"
        )
    print("PYTHON_SIMULTANEOUS_FRONTIER PASS")


if __name__ == "__main__":
    main()
